compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp

```diff
@@ -0,0 +1,842 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements lowering from the AMDAIE dialect to AIE and AIEX
+// dialects.
+//
+//===----------------------------------------------------------------------===//
+
+#include "aie/Dialect/AIE/IR/AIEDialect.h"
+#include "aie/Dialect/AIEX/IR/AIEXDialect.h"
+#include "iree-amd-aie/IR/AMDAIEDialect.h"
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree-amd-aie/Transforms/Transforms.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/Iterators.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-amdaie-lower-to-aie"
+
+using namespace xilinx;
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+/// Utility to remap the provided operation's operands.
+void remapOperands(Operation *op, IRMapping &mapper) {
+  for (int i = 0; i < op->getNumOperands(); ++i) {
+    Value operand = op->getOperand(i);
+    if (mapper.contains(operand)) {
+      op->setOperand(i, mapper.lookup(operand));
+    }
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Convert amdaie.core operation to aie.core
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+void convertSizeStrideToBDDimLayoutAttr(
+    IRRewriter &rewriter, const SmallVector<OpFoldResult> &sizes,
+    const SmallVector<OpFoldResult> &strides,
+    SmallVectorImpl<AIE::BDDimLayoutAttr> &bdDimLayoutAttr) {
+  bdDimLayoutAttr.reserve(sizes.size());
+  for (auto [size, stride] : llvm::zip(sizes, strides)) {
+    bdDimLayoutAttr.push_back(AIE::BDDimLayoutAttr::get(
+        rewriter.getContext(), getConstantIntValue(size).value(),
+        getConstantIntValue(stride).value()));
+  }
+}
+
+/// Utility to create an `aie.objectfifo` operation from
+/// `amdaie.circular_dma_cpy_nd`.
+AIE::ObjectFifoCreateOp createObjectFifo(IRRewriter &rewriter,
+                                         AMDAIE::CircularDmaCpyNdOp dmaOp,
+                                         Value srcTile, ValueRange dstTiles,
+                                         StringAttr &symName) {
+  // Create BD layout attributes from DMA `strides` and `sizes`.
+  SmallVector<AIE::BDDimLayoutAttr, 4> srcBDDimLayoutAttrs;
+  convertSizeStrideToBDDimLayoutAttr(rewriter, dmaOp.getSourceMixedSizes(),
+                                     dmaOp.getSourceMixedStrides(),
+                                     srcBDDimLayoutAttrs);
+  SmallVector<AIE::BDDimLayoutAttr, 4> dstBDDimLayoutAttrs;
+  convertSizeStrideToBDDimLayoutAttr(rewriter, dmaOp.getTargetMixedSizes(),
+                                     dmaOp.getTargetMixedStrides(),
+                                     dstBDDimLayoutAttrs);
+
+  AIE::BDDimLayoutArrayAttr srcDims = AIE::BDDimLayoutArrayAttr::get(
+      rewriter.getContext(), ArrayRef(srcBDDimLayoutAttrs));
```

Comment:
I think you can also put this inside the same utility function? It makes the codes look cleaner.